packages/world/src/utils/InventoryUtils.sol

```diff
@@ -1,244 +1,404 @@
+
+import { ObjectAmount, ObjectTypeLib } from "../ObjectTypeLib.sol";
+import { ObjectTypes } from "../ObjectTypes.sol";
 import { getUniqueEntity } from "../Utils.sol";
 
 import { EntityId } from "../EntityId.sol";
 
 using ObjectTypeLib for ObjectTypeId;
 
-function addToInventory(
-  EntityId owner,
-  ObjectTypeId ownerObjectTypeId,
-  ObjectTypeId objectTypeId,
-  uint16 numObjectsToAdd
-) {
-  require(owner.exists(), "Owner entity does not exist");
-  uint16 stackable = ObjectTypeMetadata._getStackable(objectTypeId);
-  require(stackable > 0, "This object type cannot be added to the inventory");
-
-  uint16 numInitialObjects = InventoryCount._get(owner, objectTypeId);
-  uint16 numInitialFullStacks = numInitialObjects / stackable;
-  bool hasInitialPartialStack = numInitialObjects % stackable != 0;
-  uint16 numFinalObjects = numInitialObjects + numObjectsToAdd;
-  uint16 numFinalFullStacks = numFinalObjects / stackable;
-  bool hasFinalPartialStack = numFinalObjects % stackable != 0;
-
-  uint16 numInitialSlotsUsed = InventorySlots._get(owner);
-  uint16 numFinalSlotsUsedDelta =
-    (numFinalFullStacks + (hasFinalPartialStack ? 1 : 0)) - (numInitialFullStacks + (hasInitialPartialStack ? 1 : 0));
-  uint16 numFinalSlotsUsed = numInitialSlotsUsed + numFinalSlotsUsedDelta;
-  require(numFinalSlotsUsed <= ObjectTypeMetadata._getMaxInventorySlots(ownerObjectTypeId), "Inventory is full");
-  InventorySlots._set(owner, numFinalSlotsUsed);
-  InventoryCount._set(owner, objectTypeId, numFinalObjects);
-
-  if (numInitialObjects == 0) {
-    InventoryObjects._push(owner, objectTypeId.unwrap());
-  }
+struct SlotTransfer {
+  uint16 slotFrom;
+  uint16 slotTo;
+  uint16 amount; // For entities, this should always be 1
 }
 
-function removeFromInventory(EntityId owner, ObjectTypeId objectTypeId, uint16 numObjectsToRemove) {
-  uint16 numInitialObjects = InventoryCount._get(owner, objectTypeId);
-  require(numInitialObjects >= numObjectsToRemove, "Not enough objects in the inventory");
+library InventoryUtils {
+  function useTool(EntityId owner, uint16 slot, uint128 useMassMax)
+    public
+    returns (uint128 massUsed, ObjectTypeId toolType)
+  {
+    EntityId tool = InventorySlot._getEntityId(owner, slot);
+    if (!tool.exists()) {
+      return (0, ObjectTypes.Null);
+    }
 
-  uint16 stackable = ObjectTypeMetadata._getStackable(objectTypeId);
-  require(stackable > 0, "This object type cannot be removed from the inventory");
+    toolType = ObjectType._get(tool);
+    require(toolType.isTool(), "Inventory item is not a tool");
+    uint128 toolMassLeft = Mass._getMass(tool);
+    require(toolMassLeft > 0, "Tool is already broken");
 
-  uint16 numInitialFullStacks = numInitialObjects / stackable;
-  bool hasInitialPartialStack = numInitialObjects % stackable != 0;
+    uint128 maxMass = ObjectTypeMetadata._getMass(toolType);
+    uint128 maxUsePerCall = maxMass / 10; // Limit to 10% of max mass per use
```

Comment:
niceee, so this fixes https://github.com/dustproject/dust/issues/149

Reply:
yes! wanted to confirm if that is the exact thing we want

Reply:
yup thats how I was thinking of it